homeassistant/components/sensor/modbus.py

```diff
@@ -1,100 +1,80 @@
+    CONF_NAME, CONF_OFFSET, CONF_UNIT_OF_MEASUREMENT)
 from homeassistant.helpers.entity import Entity
+from homeassistant.helpers import config_validation as cv
+from homeassistant.components.sensor import PLATFORM_SCHEMA
 
 _LOGGER = logging.getLogger(__name__)
 DEPENDENCIES = ['modbus']
 
+CONF_COUNT = "count"
+CONF_PRECISION = "precision"
+CONF_REGISTER = "register"
+CONF_REGISTERS = "registers"
+CONF_SCALE = "scale"
+CONF_SLAVE = "slave"
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_REGISTERS): [{
+        vol.Required(CONF_NAME): cv.string,
+        vol.Required(CONF_REGISTER): cv.positive_int,
+        vol.Optional(CONF_COUNT): cv.positive_int,
+        vol.Optional(CONF_OFFSET): vol.Coerce(float),
+        vol.Optional(CONF_PRECISION): cv.positive_int,
+        vol.Optional(CONF_SCALE): vol.Coerce(float),
+        vol.Optional(CONF_SLAVE): cv.positive_int,
+        vol.Optional(CONF_UNIT_OF_MEASUREMENT): cv.string
+    }]
+})
+
 
 def setup_platform(hass, config, add_devices, discovery_info=None):
-    """Setup Modbus devices."""
+    """Setup Modbus sensors."""
     sensors = []
-    slave = config.get("slave", None)
-    if modbus.TYPE == "serial" and not slave:
-        _LOGGER.error("No slave number provided for serial Modbus")
-        return False
-    registers = config.get("registers")
-    if registers:
-        for regnum, register in registers.items():
-            if register.get("name"):
-                sensors.append(
-                    ModbusSensor(register.get("name"),
-                                 slave,
-                                 regnum,
-                                 None,
-                                 register.get("unit"),
-                                 scale=register.get("scale", 1),
-                                 offset=register.get("offset", 0),
-                                 precision=register.get("precision", 0)))
-            if register.get("bits"):
-                bits = register.get("bits")
-                for bitnum, bit in bits.items():
-                    if bit.get("name"):
-                        sensors.append(ModbusSensor(bit.get("name"),
-                                                    slave,
-                                                    regnum,
-                                                    bitnum))
-    coils = config.get("coils")
-    if coils:
-        for coilnum, coil in coils.items():
-            sensors.append(ModbusSensor(coil.get("name"),
-                                        slave,
-                                        coilnum,
-                                        coil=True))
-
+    for register in config.get(CONF_REGISTERS):
+        sensors.append(ModbusRegisterSensor(
+            register.get(CONF_NAME),
+            register.get(CONF_SLAVE),
+            register.get(CONF_REGISTER),
+            register.get(CONF_UNIT_OF_MEASUREMENT),
+            register.get(CONF_COUNT, 1),
```

Comment:
Maybe you can include some of these defaults in the Schema: `vol.Optional(CONF_COUNT, default=1): cv.positive_int`